crates/core/tedge_agent/src/restart_manager/actor.rs

```diff
@@ -0,0 +1,190 @@
+use crate::restart_manager::config::RestartManagerConfig;
+use crate::restart_manager::error::RestartManagerError;
+use crate::restart_manager::restart_operation_handler::restart_operation::create_tmp_restart_file;
+use crate::restart_manager::restart_operation_handler::restart_operation::has_rebooted;
+use crate::state_repository::state::AgentStateRepository;
+use crate::state_repository::state::RestartOperationStatus;
+use crate::state_repository::state::State;
+use crate::state_repository::state::StateRepository;
+use crate::state_repository::state::StateStatus;
+use async_trait::async_trait;
+use std::time::Duration;
+use tedge_actors::Actor;
+use tedge_actors::MessageReceiver;
+use tedge_actors::RuntimeError;
+use tedge_actors::RuntimeRequest;
+use tedge_actors::Sender;
+use tedge_actors::SimpleMessageBox;
+use tedge_actors::WrappedInput;
+use tedge_api::OperationStatus;
+use tedge_api::RestartOperationRequest;
+use tedge_api::RestartOperationResponse;
+use tedge_config::system_services::SystemConfig;
+use tokio::process::Command;
+use tokio::time::timeout;
+use tracing::error;
+use tracing::info;
+
+const SYNC: &str = "sync";
+
+#[cfg(not(test))]
+const SUDO: &str = "sudo";
+#[cfg(test)]
+const SUDO: &str = "echo";
+
+pub struct RestartManagerActor {
+    config: RestartManagerConfig,
+    state_repository: AgentStateRepository,
+    message_box: SimpleMessageBox<RestartOperationRequest, RestartOperationResponse>,
+}
+
+#[async_trait]
+impl Actor for RestartManagerActor {
+    fn name(&self) -> &str {
+        "RestartManagerActor"
+    }
+
+    async fn run(&mut self) -> Result<(), RuntimeError> {
+        self.process_pending_restart_operation().await?;
+
+        while let Some(request) = self.message_box.recv().await {
+            let maybe_error = self.handle_restart_operation(&request).await;
+
+            match timeout(Duration::from_secs(5), self.message_box.recv_message()).await {
```

Comment:
[follow-up task] We need to specifically await for a shutdown message without consuming any request if the shutdown fail. Doing will remove the case where a second request is received before the shutdown is handled.

Implementing this needs to update the whole hierarchy of receiver with a `recv_signal()` method.